src/runtime/recipe/type-checker.ts

```diff
@@ -76,7 +82,7 @@ export class TypeChecker {
         }
         return candidate;
       }
-      return null;
+      return error(`could not guarantee variable ${candidate} meets read requirements ${candidate.canReadSubset} with write guarantees ${candidate.canWriteSuperset}`);
```

Comment:
why does the containing function here (processTypeList) return null sometimes and throw an error sometimes now? If there's a good rationale, make it clear in a comment.

Reply:
Sorry, where is the throw?

This is returning null via a helper function that adds an error message to options' error map.
Maybe I should rename `error` to `returnWithMessage` or something similar?

Reply:
As the error function is currently only used in this one place, I've inlined it.

We can generalize the pattern when necessary.